packages/server/src/time-control/time-handler-sequential.ts

```diff
@@ -60,58 +64,60 @@ export class TimeHandlerSequentialMoves implements ITimeHandler {
     playerNr: number,
   ): ITimeControlBase {
     const config = game.config as IConfigWithTimeControl;
+    const timestamp = this._clock.getTimestamp();
 
-    switch (config.time_control.type) {
-      case TimeControlType.Absolute: {
-        let timeControl = game.time_control;
-        // could happen for old games
-        if (timeControl === undefined) {
-          timeControl = this.initialState(game.variant, config);
-        }
-        const playerData = timeControl.forPlayer[playerNr];
-        const timestamp = new Date();
+    // could happen for old games
+    // TODO: remove? maybe after db migration?
+    if (game.time_control === undefined) {
+      game.time_control = this.initialState(game.variant, config);
+    }
 
-        timeControl.moveTimestamps.push(timestamp);
+    // mutates its input
+    let transition: (playerTimeControl: IPerPlayerTimeControlBase) => void;
 
-        if (playerData.onThePlaySince !== null) {
+    switch (config.time_control.type) {
+      case TimeControlType.Absolute: {
+        transition = (playerData) => {
           playerData.remainingTimeMS -=
             timestamp.getTime() - playerData.onThePlaySince.getTime();
-        }
-        playerData.onThePlaySince = null;
-        this._timeoutService.clearPlayerTimeout(game.id, playerNr);
-
-        // time control starts only after the first move of player
-        const nextPlayers = game_obj
-          .nextToPlay()
-          .filter(
-            (player) =>
-              game.moves.some((move) => player in move) || player === playerNr,
-          );
-        nextPlayers.forEach((player) => {
-          timeControl.forPlayer[player].onThePlaySince = timestamp;
-          this._timeoutService.scheduleTimeout(
-            game.id,
-            player,
-            timeControl.forPlayer[player].remainingTimeMS,
-          );
-        });
-
-        return timeControl;
+        };
+        break;
+      }
+
+      case TimeControlType.Fischer: {
+        // TODO: how to avoid type cast, typeguard?
```

Comment:
I've been thinking, maybe I could change the interface of ITimeHandler s.t. it takes 1-2 generic types (config, and the type for stored data). This would not get rid of the type assertion, but maybe make it easier for developers to get the types correct.

The idea is that, as long as initialState creates the correct type, and other code parts don't mutate the time control stuff in db, then the type assertion should be ok.